R/api_calls.R

```diff
@@ -57,7 +79,93 @@ See ?ed_auth.")
 Contact technology@eidith.org.")
   }
 
-  if(verbose) message("Importing...")
+  if(verbose) cat_line("Importing...\n")
+  data <- fromJSON(content(request, as = "text", encoding="UTF-8"))
+
+  if(header_only) {
+    return(data)
+  } else {
+    data <- as_tibble(data)
+  }
+
+  # if(nrow(data) == 0){
+  #   message((paste0("Download for the ", endpoint, " table failed. Please contact technology@eidith.org")))
+  #   return(invisible(0))
+  # }
+
+  if("ExceptionMessage" %in% names(data)){
+    cat_line(red(paste0("Download for the ", endpoint, " table failed. Please contact technology@eidith.org\n")))
+    return(invisible(0))
+  }
+
+  if(postprocess){
+    data <- tryCatch(ed_process(data, endpoint),
+                     error = function(e){
+                       cat_line(red(paste0("Error: There are unexpected fields in the ", endpoint, " download. Please contact technology@eidith.org.\n")))
+                       return(invisible(NULL))
+                     })
+
+}
+
+  return(data)
+}
+
+
+#' @param verbose Show a progress bar and other messages?
+#' @param header_only Return only the table header.  Useful for checking if
+#' API access works
+#' @param lmdate_from filter records by earliest date, in YYYY-MM-DD format
+#' @param lmdate_to filter records by latest date, in YYYY-MM-DD format (see details)
+#' @param ... additional arguments passed to [httr::GET()]
+#' @return a [tibble][tibble::tibble()]-style data frame
+#' @rdname ed2_get
+#' @name ed2_get
+
+
+#' @noRd
+#' @importFrom httr GET status_code progress authenticate content modify_url
+#' @importFrom jsonlite fromJSON
+#' @importFrom tibble as_tibble
+#' @export
+ed2_get <- function(endpoint2, postprocess=TRUE, verbose=interactive(),
+                   header_only=FALSE, lmdate_from="2000-01-01",
+                   lmdate_to=Sys.Date() + 1, auth=NULL, ...) {
+
+  if(endpoint2 == "TestDataInterpreted" | endpoint2 == "TestDataSerology"){
+    url <- modify_url(url = paste0(eidith2_base_url, "Extract", endpoint2),
+                      query = list(header_only = ifelse(header_only, "y", "n"),
+                                   lmdate_from = lmdate_from,
+                                   lmdate_to = lmdate_to))
+  }else{
+  url <- modify_url(url = paste0(eidith2_base_url, "Extract", endpoint2, "Data"),
+                    query = list(header_only = ifelse(header_only, "y", "n"),
+                                 lmdate_from = lmdate_from,
+                                 lmdate_to = lmdate_to))
+  }
+
+  if(verbose) {
+    pbar <- progress()
+    cat_line(red(paste("Downloading PREDICT-2", endpoint2, "table...\n")))
+  } else {
+    pbar <- NULL
+  }
+
+  if(is.null(auth)) auth <- ed_auth()
+
+  request <- GET(url=url, authenticate(auth[1], auth[2], type="basic"),
+                 pbar, ...)
+
+  if(status_code(request) == 401) {
+    stop("Unauthorized (HTTP 401). Your username or password do not match an account.
+         See ?ed_auth.")
+  }
```

Comment:
We should put the `usethis::edit_r_environ()` in the docs for `ed_auth()`

Reply:
Added.